server/commands/connected_objects.go

```diff
@@ -0,0 +1,351 @@
+			err := c.reverseExpandTupleToUserset(ctx, &reverseExpandRequest{
+				storeID:          storeID,
+				ingress:          ingress,
+				sourceObjectRef:  sourceObjRef,
+				targetUserRef:    req.User,
+				contextualTuples: req.ContextualTuples,
+				limit:            req.Limit,
+			}, resultChan, foundObjectsMap, foundCount)
+			if err != nil {
+				return err
+			}
+
+		default:
+			return errors.Errorf("unsupported ingress type")
+		}
+	}
+
+	return nil
+}
+
+// StreamedConnectedObjects yields all of the objects of the provided objectType that
+// the given user has a specific relation with. The results will be limited by the request
+// limit. If a 0 limit is provided then all objects of the provided objectType will be
+// returned.
+func (c *ConnectedObjectsCommand) StreamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string, // object string (e.g. document:1)
+) error {
+
+	var foundCount *uint32
+	if req.Limit > 0 {
+		foundCount = new(uint32)
+	}
+
+	var foundObjects sync.Map
+	return c.streamedConnectedObjects(ctx, req, resultChan, &foundObjects, foundCount)
+}
+
+type reverseExpandRequest struct {
+	storeID          string
+	ingress          *graph.RelationshipIngress
+	sourceObjectRef  *openfgapb.RelationReference
+	targetUserRef    *openfgapb.ObjectRelation
+	contextualTuples []*openfgapb.TupleKey
+	limit            uint32
+}
+
+func (c *ConnectedObjectsCommand) reverseExpandTupleToUserset(
+	ctx context.Context,
+	req *reverseExpandRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	store := req.storeID
+
+	ingress := req.ingress.Ingress
+
+	sourceObjectType := req.sourceObjectRef.GetType()
+	sourceObjectRel := req.sourceObjectRef.GetRelation()
+
+	tuplesetRelation := req.ingress.TuplesetRelation.GetRelation()
+
+	var tuples []*openfgapb.Tuple
+	for _, t := range req.contextualTuples {
+
+		object := t.GetObject()
+		objectType, _ := tuple.SplitObject(object)
+		if objectType != ingress.GetType() {
+			continue
+		}
+
+		if t.GetRelation() != tuplesetRelation {
+			continue
+		}
+
+		targetUserStr := req.targetUserRef.GetObject()
+
+		userObj, _ := tuple.SplitObjectRelation(t.GetUser())
+		if userObj == targetUserStr || userObj == "*" {
+			tuples = append(tuples, &openfgapb.Tuple{Key: t})
+		}
+	}
+	iter1 := storage.NewStaticTupleIterator(tuples)
+
+	iter2, err := c.Datastore.ReadStartingWithUser(ctx, store, storage.ReadStartingWithUserFilter{
+		ObjectType: req.ingress.Ingress.GetType(),
+		Relation:   tuplesetRelation,
+		UserFilter: []*openfgapb.ObjectRelation{
+			{Object: req.targetUserRef.Object},
+			{Object: "*"},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	iter := storage.NewCombinedIterator(iter1, iter2)
+
+	g := errgroup.Group{}
+	g.SetLimit(100) // set some concurrency limit
+
+	for {
+		t, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
```

Comment:
Missing handling when the error isn't the iterator has been consumed.

Reply:
Resolved.